lib/apiservers/engine/backends/container.go

```diff
@@ -689,8 +694,12 @@ func (c *Container) ContainerRm(name string, config *types.ContainerRmConfig) er
 	if err != nil {
 		switch err := err.(type) {
 		case *containers.ContainerRemoveNotFound:
-			cache.ContainerCache().DeleteContainer(id)
-			return NotFoundError(name)
+			// if this is an autoremove container then it was deleted by the die event
+			// and will be removed from the cache on the destroy event
+			if !vc.HostConfig.AutoRemove {
+				cache.ContainerCache().DeleteContainer(id)
```

Comment:
Yes it could happen...I don't know that we can make remove idempotent...but a better way to handle here would be to always call the cache delete / portmap clean up and then conditionally return the error